ssl/ssl_sess.c

```diff
@@ -619,16 +624,23 @@ int ssl_get_prev_session(SSL *s, CLIENTHELLO_MSG *hello)
         goto err;
     }
 
-    s->session_ctx->stats.sess_hit++;
+    if (!SSL_IS_TLS13(s)) {
+        /* We already did this for TLS1.3 */
+        SSL_SESSION_free(s->session);
+        s->session = ret;
+    }
 
-    SSL_SESSION_free(s->session);
-    s->session = ret;
+    s->session_ctx->stats.sess_hit++;
     s->verify_result = s->session->verify_result;
+
     return 1;
 
  err:
     if (ret != NULL) {
         SSL_SESSION_free(ret);
+        /* In TLSv1.3 we already set s->session, so better NULL it out */
```

Comment:
`s->session` has the same value as `ret` which we just freed, so the value of `s->session` is now invalid - hence we NULL it out. That was supposed to be what the comment was there to explain, but on reading it again perhaps it wasn't so clear. I have rephrased the comment.